Sources/JSONUtilities/Dictionary+JSONKey.swift

```diff
@@ -174,159 +126,172 @@ extension Dictionary where Key: StringProtocol {
-      return nil
-    }
-
-    return T.from(jsonValue: jsonValue)
+    return try? json(atKeyPath: keyPath)
   }
 
   // MARK: [JSONPrimitiveConvertible] type
 
   /// Decode an array of custom raw types with a mandatory key
-  public func json<T: JSONPrimitiveConvertible>(atKeyPath keyPath: Key) throws -> [T] {
-
-    guard let jsonValues = self[keyPath: keyPath] as? [T.JSONType] else {
-      throw DecodingError.mandatoryKeyNotFound(key: keyPath)
-    }
+  public func json<T: JSONPrimitiveConvertible>(atKeyPath keyPath: Key, invalidItemBehaviour: InvalidItemBehaviour = .remove) throws -> [T] {
+    return try decodeArray(atKeyPath: keyPath, invalidItemBehaviour: invalidItemBehaviour) { jsonArray, value in
+      let jsonValue: T.JSONType = try getValue(array: jsonArray, value: value)
 
-    return jsonValues.flatMap {
-      T.from(jsonValue: $0)
+      guard let transformedValue = T.from(jsonValue: jsonValue) else {
+        throw DecodingError.conversionFailure(type: T.self, value: jsonValue)
+      }
+      return transformedValue
     }
-
   }
 
   /// Optionally decode an array custom raw types with a mandatory key
-  public func json<T: JSONPrimitiveConvertible>(atKeyPath keyPath: Key) -> [T]? {
+  public func json<T: JSONPrimitiveConvertible>(atKeyPath keyPath: Key, invalidItemBehaviour: InvalidItemBehaviour = .remove) -> [T]? {
+    return try? json(atKeyPath: keyPath, invalidItemBehaviour: invalidItemBehaviour)
+  }
 
-    guard let jsonValues = self[keyPath: keyPath] as? [T.JSONType] else {
-      return nil
-    }
+  // MARK: JSONDictionary and JSONArray creation
 
-    return jsonValues.flatMap {
-      T.from(jsonValue:$0)
-    }
+  fileprivate func JSONDictionaryForKey(atKeyPath keyPath: Key) throws -> JSONDictionary {
+    return try getValue(atKeyPath: keyPath)
+  }
 
+  fileprivate func JSONArrayForKey(atKeyPath keyPath: Key) throws -> JSONArray {
+    return try getValue(atKeyPath: keyPath)
   }
 
-  // MARK: JSONDictionary and JSONArray creation
+  // MARK: Value decoding
 
-  fileprivate func JSONDictionaryForKey(atKeyPath keyPath: Key) throws -> JSONDictionary {
-    guard let jsonDictionary = self[keyPath: keyPath] as? JSONDictionary else {
-      throw DecodingError.mandatoryKeyNotFound(key: keyPath)
+  fileprivate func getValue<A, B>(array: [A], value: A) throws -> B {
+    guard let typedValue = value as? B else {
+      throw DecodingError.incorrectTypeInArray(array: array, expectedType: B.self, value: value)
     }
-    return jsonDictionary
+    return typedValue
   }
 
-  fileprivate func JSONArrayForKey(atKeyPath keyPath: Key) throws -> JSONArray {
-    guard let jsonArray = self[keyPath: keyPath] as? JSONArray else {
-      throw DecodingError.mandatoryKeyNotFound(key: keyPath)
+  fileprivate func getValue<T>(atKeyPath key: Key) throws -> T {
+    guard let value = self[key] else {
+      // swiftlint:disable:next force_cast
+      throw DecodingError.keyNotFound(dictionary: self as! JSONDictionary, key: key)
```

Comment:
Maybe for now `keyNotFound` should take `Any` to avoid having to force cast `self` to a `JSONDictionary` as this fails on older versions of Xcode as currently set in TravisCI, not sure how quickly people get into the latest version, so maybe we can give some time until we do this.